src/dune/mdx.ml

```diff
@@ -0,0 +1,247 @@
+    in
+    fun dir -> Dep.file_selector (File_selector.create ~dir pred)
+
+  let source_tree_dep_set dir =
+    let prefix_with, dir = Path.extract_build_context_dir_exn dir in
+    match File_tree.find_dir dir with
+    | None -> Dep.Set.empty
+    | Some dir ->
+      File_tree.Dir.fold dir ~init:Dep.Set.empty
+        ~traverse:Sub_dirs.Status.Set.all ~f:(fun dir acc ->
+          let files = File_tree.Dir.files dir in
+          let path = Path.append_source prefix_with (File_tree.Dir.path dir) in
+          match String.Set.is_empty files with
+          | true -> Dep.Set.add acc (dir_without_files_dep path)
+          | false ->
+            let paths =
+              String.Set.fold files ~init:Path.Set.empty ~f:(fun fn acc ->
+                  Path.Set.add acc (Path.relative path fn))
+            in
+            Dep.Set.add_paths acc paths)
+
+  let to_dep_set ~dir t_list =
+    let dirs, files = dirs_and_files ~dir t_list in
+    let dep_set = Dep.Set.of_files files in
+    List.fold_left dirs ~init:dep_set ~f:(fun acc dir ->
+        Dep.Set.union acc (source_tree_dep_set dir))
+end
+
+module Prelude = struct
+  type t =
+    | Default of Path.Local.t
+    | Env of
+        { env : string
+        ; file : Path.Local.t
+        }
+
+  let decode =
+    let open Dune_lang.Decoder in
+    let path = string >>| Path.Local.of_string in
+    let decode_env =
+      let+ () = keyword "env"
+      and+ env = string
+      and+ file = path in
+      Env { env; file }
+    in
+    let decode_default =
+      let+ file = path in
+      Default file
+    in
+    if_list ~then_:(enter decode_env) ~else_:decode_default
+
+  let to_args ~dir t : _ Command.Args.t list =
+    let bpath p = Path.build (Path.Build.append_local dir p) in
+    match t with
+    | Default file -> [ A "--prelude"; Dep (bpath file) ]
+    | Env {env; file} ->
+      let arg = Printf.sprintf "%s:%s" env (Path.Local.to_string file) in
+      [ A "--prelude"; A arg; Hidden_deps (Dep.Set.of_files [bpath file]) ]
+end
+
+type t =
+  { loc : Loc.t
+  ; files : Predicate_lang.Glob.t
+  ; packages : Package.Name.t list
+  ; preludes : Prelude.t list
+  }
+
+type Stanza.t += T of t
+
+let syntax =
+  let name = "mdx" in
+  let desc = "mdx extension to verify code blocks in .md and .mli" in
+  Dune_lang.Syntax.create ~name ~desc [ (0, 1) ]
+
+let default_files =
+  let has_extention ext s = String.equal ext (Filename.extension s) in
+  let md_files = Predicate_lang.Glob.of_pred (has_extention ".md") in
+  let mli_files = Predicate_lang.Glob.of_pred (has_extention ".mli") in
+  Predicate_lang.union [ md_files; mli_files ]
+
+let decode =
+  let open Dune_lang.Decoder in
+  fields
+    (let+ loc = loc
+     and+ files =
+       field "files" Predicate_lang.Glob.decode ~default:default_files
+     and+ packages = field_o "packages" (repeat Package.Name.decode)
+     and+ preludes = field_o "preludes" (repeat Prelude.decode) in
+     let packages = Option.value ~default:[] packages in
```

Comment:
I believe you can use `field ~default:[]` to simplify the definition of packages and preludes.